### PR TITLE
SALTO-1715: addDeploymentAnnotations now supports inner types

### PR DESCRIPTION
**Pull request by alonstern**
Added support for inner types in `addDeploymentAnnotations`

---
_Release Notes_: 
None

---
_User Notifications_: 
None

### CODE REVIEW

File: packages/adapter-components/src/elements/swagger/deployment/annotations.ts [hunk 1]
```diff
@@ -13,49 +13,84 @@
...
+  if (annotatedTypes.has(type.elemID.getFullName())) {
+    return
+  }
+
+  annotatedTypes.add(type.elemID.getFullName())
+
+  const fields = getFields(
+    swagger,
+    schema
+  )
```
Comment (netama) [inline packages/adapter-components/src/elements/swagger/deployment/annotations.ts:64]: nit:
```suggestion
  const fields = getFields(swagger, schema)
```

File: packages/adapter-components/src/elements/swagger/deployment/annotations.ts [hunk 2]
```diff
@@ -13,49 +13,84 @@
...
+
+  await awu(Object.entries(fields)).forEach(async ([name, properties]) => {
+    const field = type.fields[name]
+
+    if (field === undefined) {
+      return
+    }
+    field.annotations[OPERATION_TO_ANNOTATION[action]] = true
```
Comment (netama) [inline packages/adapter-components/src/elements/swagger/deployment/annotations.ts:72]: isn't true the default?

File: packages/adapter-components/src/elements/swagger/deployment/annotations.ts [hunk 3]
```diff
@@ -13,49 +13,84 @@
...
+
+    const fieldType = await field.getType()
+
+    if (isObjectType(fieldType)) {
+      await setTypeAnnotations(fieldType, properties, swagger, action, annotatedTypes)
+    }
+
+    if (isListType(fieldType)) {
```
Comment (netama) [inline packages/adapter-components/src/elements/swagger/deployment/annotations.ts:80]: what about map types? and specifically `additionalProperties`?

File: packages/adapter-components/src/elements/swagger/deployment/annotations.ts [hunk 4]
```diff
@@ -76,19 +111,17 @@ export const addDeploymentAnnotationsFromSwagger = (
       return
     }
 
-    delete type.annotations[OPERATION_TO_ANNOTATION[operation as ActionName]]
+    type.annotations[OPERATION_TO_ANNOTATION[operation as ActionName]] = true
```
Comment (netama) [inline packages/adapter-components/src/elements/swagger/deployment/annotations.ts:114]: why the change?

File: packages/adapter-components/src/elements/swagger/deployment/annotations.ts [hunk 5]
```diff
@@ -76,19 +111,17 @@ export const addDeploymentAnnotationsFromSwagger = (
...
     )
 
-    fields.forEach(fieldName => {
-      const field = type.fields[fieldName]
-      if (field !== undefined) {
-        delete field.annotations[OPERATION_TO_ANNOTATION[operation as ActionName]]
-      }
-    })
+    if (schema === undefined) {
+      return
```
Comment (netama) [inline packages/adapter-components/src/elements/swagger/deployment/annotations.ts:122]: should we log a warning in this case?

File: packages/adapter-components/src/elements/swagger/deployment/annotations.ts [hunk 6]
```diff
@@ -13,49 +13,84 @@
...
+  action: ActionName,
+  annotatedTypes: Set<string>
+): Promise<void> => {
+  if (annotatedTypes.has(type.elemID.getFullName())) {
+    return
+  }
+
+  annotatedTypes.add(type.elemID.getFullName())
+
+  const fields = getFields(
```
Comment (netama) [inline packages/adapter-components/src/elements/swagger/deployment/annotations.ts:61]: nit: maybe `swaggerFields` to make it easier to distinguish from the existing objecttype's fields?

File: packages/adapter-components/src/elements/swagger/deployment/annotations.ts [hunk 7]
```diff
@@ -13,49 +13,84 @@
...
+    swagger,
+    schema
+  )
+
+  await awu(Object.entries(fields)).forEach(async ([name, properties]) => {
```
Comment (netama) [inline packages/adapter-components/src/elements/swagger/deployment/annotations.ts:66]: technically we allow ourselves to add/remove fields with `fieldTypeOverrides` - do we need any handling for these?

File: packages/adapter-components/src/elements/swagger/deployment/annotations.ts [hunk 8]
```diff
@@ -100,19 +133,31 @@ export const addDeploymentAnnotationsFromSwagger = (
...
+const addDeploymentAnnotationsToType = async (
   type: ObjectType,
   swaggers: LoadedSwagger[],
   endpointDetails: DeploymentRequestsByAction,
-): void => {
-  type.annotations[DEPLOYMENT_ANNOTATIONS.CREATABLE] = false
-  type.annotations[DEPLOYMENT_ANNOTATIONS.UPDATABLE] = false
-  type.annotations[DEPLOYMENT_ANNOTATIONS.DELETABLE] = false
+): Promise<void> => {
+  type.annotations[DEPLOYMENT_ANNOTATIONS.CREATABLE] = Boolean(
```
Comment (netama) [inline packages/adapter-components/src/elements/swagger/deployment/annotations.ts:141]: why the `Boolean()` constructor?

File: packages/adapter-components/src/elements/swagger/deployment/annotations.ts [hunk 9]
```diff
@@ -100,19 +133,31 @@ export const addDeploymentAnnotationsFromSwagger = (
...
+    field.annotations[DEPLOYMENT_ANNOTATIONS.CREATABLE] = Boolean(
+      field.annotations[DEPLOYMENT_ANNOTATIONS.CREATABLE]
+    )
+    field.annotations[DEPLOYMENT_ANNOTATIONS.UPDATABLE] = Boolean(
+      field.annotations[DEPLOYMENT_ANNOTATIONS.UPDATABLE]
+    )
   })
-  swaggers.forEach(swagger => addDeploymentAnnotationsFromSwagger(type, swagger, endpointDetails))
+  await Promise.all(
+    swaggers.map(swagger => addDeploymentAnnotationsFromSwagger(type, swagger, endpointDetails))
```
Comment (netama) [inline packages/adapter-components/src/elements/swagger/deployment/annotations.ts:159]: I guess this was kind of already here (sorry I missed it in the previous PR), but - 
do we ever expect multiple swaggers to have definitions for the same type?
if yes - what's the expected result? any chance of overlaps for the same field (in which case there is a race condition here that we should probably avoid)?
if no - can we enforce that?

File: packages/adapter-components/src/elements/swagger/deployment/annotations.ts [hunk 10]
```diff
@@ -122,16 +167,16 @@ const addDeploymentAnnotationsToType = (
...
  */
-export const addDeploymentAnnotations = (
+export const addDeploymentAnnotations = async (
   types: ObjectType[],
   swaggers: LoadedSwagger[],
   apiDefinitions: AdapterApiConfig,
-): void => {
-  types.forEach(
+): Promise<void> => {
+  await Promise.all(types.map(
```
Comment (netama) [inline packages/adapter-components/src/elements/swagger/deployment/annotations.ts:175]: another potential race condition here (I think) - multiple types can have the same subtype, what do we expect to happen in that case? and do we really need to parallelize (and make this non-deterministic)?

**Review by netama — COMMENTED** [review]
added some comments and questions - I think it would be better to call out the various changes happening here in the PR description, as there are some that I think are different from the previous design :)
specifically - 
1. IIUC it looks like we changed the default for deployable fields from true to false? why was that? 
2. I think there are some potential race conditions with the new `Promise.all()` calls - I'm not sure I understand if there is a real risk or not, but if there isn't I think we should enforce it
3. are we covering `additionalProperties` with this?

File: packages/adapter-components/src/elements/swagger/deployment/annotations.ts (hunk 7 above)
Comment (alonstern) [reply packages/adapter-components/src/elements/swagger/deployment/annotations.ts:66]: I don't think so. I guess if we add a new field we will need to also add its annotations

File: packages/adapter-components/src/elements/swagger/deployment/annotations.ts (hunk 9 above)
Comment (alonstern) [reply packages/adapter-components/src/elements/swagger/deployment/annotations.ts:159]: Our only example of multiple swaggers currently is in Jira, in which we don't expect the same type to appear more than once. But if so, we will sort of get the wanted outcome, i.e., all the fields that are deployable with at least one of the swaggers will be marked with true, and the rest with false

File: packages/adapter-components/src/elements/swagger/deployment/annotations.ts (hunk 10 above)
Comment (alonstern) [reply packages/adapter-components/src/elements/swagger/deployment/annotations.ts:175]: Replaced with `awu.foreach`.
If two types have the same subtype, we assume that the same fields of the subtype are deployable in both types (which currently seems correct for Jira). If that's not the case, we will need to separate the subtype to multiple subtypes

File: packages/adapter-components/src/elements/swagger/deployment/annotations.ts (hunk 2 above)
Comment (alonstern) [reply packages/adapter-components/src/elements/swagger/deployment/annotations.ts:72]: The logic of this function is first, mark everything as false, then mark with true what is deployable. so in this point field.annotations[OPERATION_TO_ANNOTATION[action]] won't be `undefined`

File: packages/adapter-components/src/elements/swagger/deployment/annotations.ts (hunk 8 above)
Comment (alonstern) [reply packages/adapter-components/src/elements/swagger/deployment/annotations.ts:141]: if the annotation was already marked `true` manually, I don't want to override it, so I change it for false to use the `Boolean`

File: packages/adapter-components/src/elements/swagger/deployment/annotations.ts (hunk 4 above)
Comment (alonstern) [reply packages/adapter-components/src/elements/swagger/deployment/annotations.ts:114]: Since it is hidden now, I thought it better to be explicit about it, but it doesn't really matter

File: packages/adapter-components/src/elements/swagger/deployment/annotations.ts (hunk 7 above)
Comment (netama) [reply packages/adapter-components/src/elements/swagger/deployment/annotations.ts:66]: will it be deployable by default in this flow?

File: packages/adapter-components/src/elements/swagger/deployment/annotations.ts (hunk 8 above)
Comment (netama) [reply packages/adapter-components/src/elements/swagger/deployment/annotations.ts:141]: but didn't we say our default is also true? I think I'm missing something around when we expect this to be defined now

File: packages/adapter-components/src/elements/swagger/deployment/annotations.ts (hunk 4 above)
Comment (alonstern) [reply packages/adapter-components/src/elements/swagger/deployment/annotations.ts:122]: Changed back

**Comment by alonstern** [comment]
> added some comments and questions - I think it would be better to call out the various changes happening here in the PR description, as there are some that I think are different from the previous design :) specifically -
> 
> 1. IIUC it looks like we changed the default for deployable fields from true to false? why was that? won't it be much more noisy (even if only in the state)?
> 2. I think there are some potential race conditions with the new `Promise.all()` calls - I'm not sure I understand if there is a real risk or not, but if there isn't I think we should enforce it
> 3. are we covering `additionalProperties` with this?

1. The default was not changed, I only changed it to always write true, false and not leave the annotation empty, but that's not really important so I changed it
2. Fixed
3. Fixed

File: packages/adapter-components/src/elements/swagger/deployment/annotations.ts (hunk 7 above)
Comment (alonstern) [reply packages/adapter-components/src/elements/swagger/deployment/annotations.ts]: If the user did not add the annotations for it and it really does not appear in the swagger, then it won't be deployable (because it really isn't)

File: packages/adapter-components/src/elements/swagger/deployment/annotations.ts (hunk 8 above)
Comment (alonstern) [reply packages/adapter-components/src/elements/swagger/deployment/annotations.ts]: It's not that this function assumes that the annotation is false if it is undefined, it will still go to the swagger and if it finds that it is deployable it will remove the annotation with the false value. It's just that if the annotation were explicitly marked as true before this function ran, we don't want to override it with false (for cases that the field does not appear in the swagger, but we do have custom logic for deploying it).

The default value when the annotation is undefined is set to true by the change validator and by the `deployChange` func.
It is meant mostly for adapters that don't use this function and don't set the annotations at all or only partially, so their current behavior won't change

File: packages/adapter-components/src/elements/swagger/deployment/annotations.ts [hunk 11]
```diff
@@ -13,49 +13,92 @@
...
+    delete field.annotations[OPERATION_TO_ANNOTATION[action]]
+
+    const fieldType = await field.getType()
+
+    if (isObjectType(fieldType)) {
+      await setTypeAnnotations(fieldType, properties, swagger, action, annotatedTypes)
+    }
+
+    if (isMapType(fieldType)) {
+      const innerType = await fieldType.getInnerType()
```
Comment (netama) [inline packages/adapter-components/src/elements/swagger/deployment/annotations.ts:82]: I think both here and below this might need to be `getDeepInnerType` (for maps-of-lists etc)

**Review by netama — APPROVED** [review]
LGTM with one last nit
